third_party/move/move-compiler-v2/src/pipeline/uninitialized_use_checker.rs

```diff
@@ -0,0 +1,381 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! Implements a checker which verifies that all locals are initialized before any use.
+//! This checker does not require any other analysis to be run before.
+//! As a side effect of running this checker, function targets are annotated with
+//! the initialized state (yes, no, maybe) of all locals at each program point.
+//!
+//! There are two parts to this checker:
+//! * `InitializedStateAnalysis` which computes the initialized state of all locals at each
+//!   program point via a forward dataflow analysis.
+//! * `UninitializedUseChecker` which checks that all locals are initialized before use.
+
+use im::Vector;
+use move_binary_format::file_format::CodeOffset;
+use move_model::{ast::TempIndex, model::FunctionEnv};
+use move_stackless_bytecode::{
+    dataflow_analysis::{DataflowAnalysis, TransferFunctions},
+    dataflow_domains::{AbstractDomain, JoinResult},
+    function_target::{FunctionData, FunctionTarget},
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    stackless_bytecode::Bytecode,
+    stackless_control_flow_graph::StacklessControlFlowGraph,
+};
+use std::collections::BTreeMap;
+
+/// State of initialization of a local at a given program point.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum Initialized {
+    No,    // definitely not initialized
+    Maybe, // maybe initialized
+    Yes,   // definitely initialized
+}
+
+impl AbstractDomain for Initialized {
+    /// Implements `join` for the initialized state lattice:
+    ///
+    ///             +-------+
+    ///             | Maybe |
+    ///             +-------+
+    ///               /   \
+    ///              /     \
+    ///      +-----+/       \+----+
+    ///      | Yes |         | No |
+    ///      +-----+\       /+----+
+    ///              \     /
+    ///               \   /
+    ///            +--------+
+    ///            | bottom |
+    ///            +--------+
+    /// Note that bottom is not explicitly represented in the code.
```

Comment:
Maybe I could be direct about what "bottom" represents here: the reason why we get `Option<Initialized>` at a program point is because that program point is unreachable, and therefore, the dataflow analysis does not calculate anything at that program point. At such program points, we have `bottom` values.